lib/web_ui/lib/src/engine/text/layout_service.dart

```diff
@@ -347,6 +398,61 @@ class RangeBox {
       direction,
     );
   }
+
+  /// Returns the text position within this box's range that's closest to the
+  /// given [x] offset.
+  ///
+  /// The [x] offset is expected to be relative to the left edge of the line,
+  /// just like the coordinates of this box.
+  ui.TextPosition getPositionForX(double x) {
+    spanometer.currentSpan = span as FlatTextSpan;
+
+    // Make `x` relative to this box.
+    x -= left;
+
+    final int startIndex = start.index;
+    final int endIndex = end.indexWithoutTrailingNewlines;
+    // The resulting `cutoff` is the index of the character where the `x` offset
+    // falls. We should return the text position of either `cutoff` or
+    // `cutoff + 1` depending on which one `x` is closer to.
+    //
+    //   offset x
+    //      ↓
+    // "A B C D E F"
+    //     ↑
+    //   cutoff
+    final int cutoff = spanometer.forceBreak(
+      startIndex,
+      endIndex,
+      availableWidth: x,
+      allowEmpty: true,
+    );
+
+    if (cutoff == endIndex) {
+      return ui.TextPosition(
+        offset: cutoff,
+        affinity: ui.TextAffinity.upstream,
+      );
+    }
+
```

Comment:
final int startIndex = start.index

Reply:
Lol the `startIndex` variable is already there at the top of the function, and it has the correct value that I need :) Nice catch!